File: C++/String/21_count_palindromic_subsequences.cpp
/*
Idea: using DP to store preprocessed values,
with each iteration, checking if current elements are equal, checking for next elements
*/

#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    const long long int mod = 1000000007;
    long long int add(long long int x, long long int y) {
        return ((x % mod) + (y % mod)) % mod;
    }
    long long int sub(long long int x, long long int y) {
        return ((x % mod) - (y % mod) + mod) % mod;
    }
    long long int  countPS(string str)
    {
        long long int  N = str.length();

        // create a 2D array to store the count of palindromic subsequence
        long long int  cps[N + 1][N + 1];
        memset(cps, 0, sizeof(cps));

        // palindromic subsequence of length 1
        for (long long int i = 0; i < N; i++)
            cps[i][i] = 1;

        // check subsequence of length L is palindrome or not
        for (long long int L = 2; L <= N; L++)
        {
            for (long long int i = 0; i <= N - L; i++)
            {
                long long int  k = L + i - 1;
                if (str[i] == str[k])
                    cps[i][k] = add(cps[i][k - 1], add(cps[i + 1][k], 1)) % mod;
                else
                    cps[i][k] = add(cps[i][k - 1], sub(cps[i + 1][k], cps[i + 1][k - 1])) % mod;
            }
        }

        // return total palindromic subsequence
        return cps[0][N - 1] % mod;
    }
};

//{ Driver Code Starts.
// Driver program
int main()
{
    int t;
    cin >> t;
    while (t--)
    {
        string str;
        cin >> str;
        Solution ob;
        long long int ans = ob.countPS(str);
        cout << ans << endl;
    }
}
// } Driver Code Ends

File: Python/String/21_count_palindromic_subsequences.py
"""
Idea: using DP to store preprocessed values,
with each iteration, checking if current elements are equal, checking for next elements
"""


class Solution:

    def countPS(self, s):

        t = [[-1 for i in range(1001)] for i in range(1001)]
        # Modulus value for calculations.
        mod = 10 ** 9 + 7

        def solve(s, i, j, t):
            # Base cases: if the string has only one character or no characters.
            if i == j:
                return 1
            if i > j:
                return 0
            # If the result for the current subproblem has already been calculated, return it.

            if t[i][j] != -1:
                return t[i][j]
            # If the first and last characters of the string are the same.
            elif s[i] == s[j]:
                # Recursively calculate the result for the remaining string
                t[i][j] = 1 + solve(s, i + 1, j, t) % mod + solve(s, i, j - 1, t) % mod
                t[i][j] %= mod
                return t[i][j]
            else:
                # Recursively calculate the result for the remaining string
                t[i][j] = solve(s, i + 1, j, t) % mod + solve(s, i, j - 1, t) % mod - solve(s, i + 1, j - 1, t) % mod
                t[i][j] %= mod
                return t[i][j]

        # Starting point of the recursive function.
        return solve(s, 0, len(s) - 1, t)


import sys

sys.setrecursionlimit(10 ** 6)

if __name__ == '__main__':
    t = int(input())
    for i in range(t):
        ob = Solution()
        print(ob.countPS(input().strip()))

